staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -3014,6 +3020,305 @@ func TestValidateFieldPath(t *testing.T) {
+                - bar: bar
+            `),
+			warnings: []string{
+				`root.atomicArray[0].bar: Invalid value: "string": gotta be baz`,
+			},
+		},
+		{
+			name: "we can't ratchet an error triggered by a normal CEL expression from an uncorrelatable part of the schema if one of its parent nodes has changed",
+			schema: mustSchema(`
+                type: array
+                x-kubernetes-list-type: atomic
+                items:
+                    type: object
+                    properties:
+                        bar:
+                            type: string
+                            x-kubernetes-validations:
+                            - rule: self == "baz"
+                              message: "gotta be baz"
+            `),
+			// CEL error comes from uncorrelatable portion of the schema,
+			// but it should be ratcheted anyway because it is the descendent
+			// or an unchanged correlatable node
+			oldObj: mustUnstructured(`
+                - bar: bar
+            `),
+			newObj: mustUnstructured(`
+                - bar: bar
+                - bar: baz
+            `),
+			errors: []string{
+				`root[0].bar: Invalid value: "string": gotta be baz`,
+			},
+		},
+		{
+			name: "transition rules never ratchet for correlatable schemas",
+			schema: mustSchema(`
+                type: object
+                properties:
+                    foo:
+                        type: string
+                        x-kubernetes-validations:
+                        - rule: oldSelf != "bar" && self == "baz"
+                          message: gotta be baz
+            `),
+			oldObj: mustUnstructured(`
+                foo: bar
+            `),
+			newObj: mustUnstructured(`
+                foo: bar
+            `),
+			errors: []string{
+				`root.foo: Invalid value: "string": gotta be baz`,
+			},
+		},
+		{
+			name: "changing field path does not change ratcheting logic",
+			schema: mustSchema(`
+                type: object
+                x-kubernetes-validations:
+                - rule: self.foo == "baz"
+                  message: gotta be baz
+                  fieldPath: ".foo"
+                properties:
+                    bar:
+                        type: string
+                    foo:
+                        type: string
+            `),
+			oldObj: mustUnstructured(`
+                foo: bar
+            `),
+			// Fieldpath is on unchanged field `foo`, but since rule is on the
+			// changed parent object we still get an error
+			newObj: mustUnstructured(`
+                foo: bar
+                bar: invalid
+            `),
+			errors: []string{
+				`root.foo: Invalid value: "object": gotta be baz`,
+			},
+		},
+		// How to write test that fails cost budget analysis?
+		// {
+		// 	name: "cost budget errors are not ratcheted",
+		// 	schema: mustSchema(`
+		//         type: string
+		//         maxLength: 20000000
+		//         x-kubernetes-validations:
+		//         - rule: self == "baz"
```

Comment:
@jiahuif or @cici37 

I wanted to write a test that will trip CEL's runtime cost budget. Know a small expression guaranteed to exceed the limit?

Reply:
Please create a simple expression that can hand-calculate cost, and set a limit that the expression will exceed by a very small amount (usually 1). In this way we can ensure consistency in case a change to the cost algorithm.

Reply:
Ty